### PR TITLE
Add REX intro

### PR DESCRIPTION
**Pull request by ares**
I know I said this would be more for opensource.com, however I wanted to continue providing intros. And that felt more like a Foreman specific blog post. I'd kindly ask for review and ideally merging this faster than the last post, it's more than a month since the last one already. I'm happy to merge any language suggestions :-)

### CODE REVIEW

File: _posts/2021-02-17-introduction-to-the-remote-execution-plugin.md [hunk 1]
```diff
@@ -0,0 +1,338 @@
...
+- howto
+- beginners
+- openhab
+- puppet
+- browfield
+---
+
+In this post, we’ll see how to add a remote execution plugin to the Foreman installation 
+and look at its basic functionality. We’ll create some jobs that we execute through SSH 
+on one or more of the Foreman hosts. Then we’ll look at some real-world use cases of
```
Comment (melcorr) [inline _posts/2021-02-17-introduction-to-the-remote-execution-plugin.md]:
```suggestion
on one or more Foreman hosts. Then we’ll look at some real-world use cases of
```

File: _posts/2021-02-17-introduction-to-the-remote-execution-plugin.md [hunk 2]
```diff
@@ -0,0 +1,338 @@
...
+useful SSH jobs automation. We’ll show how to deploy puppet agent on the Foreman host and
```
Comment (melcorr) [inline _posts/2021-02-17-introduction-to-the-remote-execution-plugin.md]:
```suggestion
useful SSH jobs automation. We’ll show how to deploy the Puppet agent on the Foreman host and
```

File: _posts/2021-02-17-introduction-to-the-remote-execution-plugin.md [hunk 3]
```diff
@@ -0,0 +1,338 @@
...
+illustrate the use with home automation, running commands on SSH enabled routers.
+
+<!--more-->
+
+## Plugin installation
+
+To add the remote execution feature, a plugin for Foreman and Smart proxy is required.
```
Comment (melcorr) [inline _posts/2021-02-17-introduction-to-the-remote-execution-plugin.md]:
```suggestion
To add the remote execution feature, a plugin for Foreman and Smart Proxy is required.
```

File: _posts/2021-02-17-introduction-to-the-remote-execution-plugin.md [hunk 4]
```diff
@@ -0,0 +1,338 @@
...
+The installer can handle the installation and the initial configuration of both for you.
```
Comment (melcorr) [inline _posts/2021-02-17-introduction-to-the-remote-execution-plugin.md]:
```suggestion
The installer can handle both the installation and the initial configuration.
```

File: _posts/2021-02-17-introduction-to-the-remote-execution-plugin.md [hunk 5]
```diff
@@ -0,0 +1,338 @@
...
+The Foreman installer can be run repeatedly. In case you haven’t done manual modifications
```
Comment (melcorr) [inline _posts/2021-02-17-introduction-to-the-remote-execution-plugin.md]:
```suggestion
The Foreman installer can be run repeatedly. As long as you haven’t done manual modifications
```

File: _posts/2021-02-17-introduction-to-the-remote-execution-plugin.md [hunk 6]
```diff
@@ -0,0 +1,338 @@
...
+to your installation, such as changing configuration files yourself, it will do no harm.
+You don’t have to repeat the same parameters, their values are stored from the previous
+run, therefore you can simply just add new params. The bare minimum to install or extend
```
Comment (melcorr) [inline _posts/2021-02-17-introduction-to-the-remote-execution-plugin.md]:
```suggestion
run, therefore you can simply just add new parameters. The bare minimum to install or extend
```

File: _posts/2021-02-17-introduction-to-the-remote-execution-plugin.md [hunk 7]
```diff
@@ -0,0 +1,338 @@
...
+existing Foreman instance with the Remote Execution, run the following:
+
+```
+foreman-installer --enable-foreman-plugin-remote-execution \
+              --enable-foreman-proxy-plugin-remote-execution-ssh
+```
+
+If you plan to use CLI, install a plugin for that, so you can use REX CLI commands by
```
Comment (melcorr) [inline _posts/2021-02-17-introduction-to-the-remote-execution-plugin.md]:
```suggestion
If you plan to use CLI commands, you must install a plugin for that, so you can use REX CLI commands by
```

File: _posts/2021-02-17-introduction-to-the-remote-execution-plugin.md [hunk 8]
```diff
@@ -0,0 +1,338 @@
...
+adding `--enable-foreman-cli-remote-execution` to the command above.
+
+## Running the first job
+
+Once the installer finishes the run, you will see the new menu entries in the Foreman
+instance. Let’s try running a job on some host. Make sure the host is running SSH daemon
+and Foreman can connect to it. Navigate to some existing host detail page and click the
```
Comment (melcorr) [inline _posts/2021-02-17-introduction-to-the-remote-execution-plugin.md]:
```suggestion
and Foreman can connect to it. Navigate to an existing host detail page and click the
```

File: _posts/2021-02-17-introduction-to-the-remote-execution-plugin.md [hunk 9]
```diff
@@ -0,0 +1,338 @@
...
+*Schedule Remote Job* button. Fill in a simple bash command to the command field,
+e.g. “echo Hello world”. Click Advanced fields and specify the password for the user.
+The password will be stored only for necessary time to execute the script,
```
Comment (melcorr) [inline _posts/2021-02-17-introduction-to-the-remote-execution-plugin.md]:
```suggestion
The password is stored only for the time it takes to execute the script,
```

File: _posts/2021-02-17-introduction-to-the-remote-execution-plugin.md [hunk 10]
```diff
@@ -0,0 +1,338 @@
...
+
+It is not practical to specify the SSH password for each job. Especially if you run the
+job on multiple hosts and they don’t have the same password. The installer generates a
+SSH key pair on the Foreman Proxy host. You can distribute the generated SSH public key
+or setup your existing SSH key you already use. The default SSH key pair location is at
+* ~foreman-proxy/.ssh/id_rsa_foreman_proxy
+* ~foreman-proxy/.ssh/id_rsa_foreman_proxy.pub
+
+If you replace these files with your own key, either make sure the private key is not
+protected by a passphrase or you need to specify it for every job, similarly to how we
```
Comment (melcorr) [inline _posts/2021-02-17-introduction-to-the-remote-execution-plugin.md]:
```suggestion
protected by a passphrase or you need to specify it for every job, similar to how we
```

File: _posts/2021-02-17-introduction-to-the-remote-execution-plugin.md [hunk 11]
```diff
@@ -0,0 +1,338 @@
...
+specified the password. To distribute the generated public key, you can use the following
+command
+
+```
+ssh-copy-id -i ~foreman-proxy/.ssh/id_rsa_foreman_proxy.pub root@target-machine.example.com
+```
+
+However you can use a more elegant way and do it from Foreman directly. For that we’ll
```
Comment (melcorr) [inline _posts/2021-02-17-introduction-to-the-remote-execution-plugin.md]:
```suggestion
However, you can use a more elegant way and do it from Foreman directly. For that we’ll
```

File: _posts/2021-02-17-introduction-to-the-remote-execution-plugin.md [hunk 12]
```diff
@@ -0,0 +1,338 @@
...
+
+```
+#!/bin/bash
+mkdir -p ~/.ssh
+echo “ssh-rsa AAAAB3NzaC1yc2EAA...” > ~/.ssh/authorized_keys
+```
+
+But we can also use an existing template snippet that also takes care of the right file
+permissions and deploys all public keys that Foreman is configured to use.
+To do that, use the following template content
```
Comment (melcorr) [inline _posts/2021-02-17-introduction-to-the-remote-execution-plugin.md]:
```suggestion
To do that, use the following template content:
```

File: _posts/2021-02-17-introduction-to-the-remote-execution-plugin.md [hunk 13]
```diff
@@ -0,0 +1,338 @@
...
+Senior sysadmins can create job templates for juniors, who can run them on servers without
+having direct access to them, e.g. allowing them only to restart stuck services.
+
+## Real world use case #1 - Puppet deployment in a brownfield deployment
+
+It’s the time for the first practical use case. In the previous
+[blog post](https://theforeman.org/2021/01/updating-foreman-inventory-with-system-facts.html),
+it would be handy to automate the setup of the puppet agent. Foreman already comes
+with the snippet, that is normally used during the host provisioning to bootstrap the
+puppet agent. However if you start with Foreman and want to manage an existing
```
Comment (melcorr) [inline _posts/2021-02-17-introduction-to-the-remote-execution-plugin.md]:
```suggestion
puppet agent. However, if you start with Foreman and want to manage an existing
```

File: _posts/2021-02-17-introduction-to-the-remote-execution-plugin.md [hunk 14]
```diff
@@ -0,0 +1,338 @@
...
+infrastructure, it’s useful to deploy puppet agents to all existing hosts.
+Puppet agents then ask the puppet server deployed with Foreman for their configuration.
+Puppet configuration management goes beyond this post, but this illustrates how
+Remote Execution allows to automate this task.
+
+To deploy a puppet agent, we need to install the a package, create a configuration file
```
Comment (melcorr) [inline _posts/2021-02-17-introduction-to-the-remote-execution-plugin.md]:
```suggestion
To deploy a puppet agent, we need to install a package, create a configuration file
```

File: _posts/2021-02-17-introduction-to-the-remote-execution-plugin.md [hunk 15]
```diff
@@ -0,0 +1,338 @@
...
+Or we could reuse an existing job template, that cover many more operating systems and
+is installed by default with the plugin. This is how we can install package on all
+supported OSes in job templates:
+
+```
+<%= render_template("Package Action - SSH Default", :action => "install", :package => 'puppet') %>
+```
+
+In the same way, you could customize the puppet server or ca server based on what is
+assigned to the host. Before we go too far, I want to point out that thanks to the
```
Comment (melcorr) [inline _posts/2021-02-17-introduction-to-the-remote-execution-plugin.md]:
```suggestion
assigned to the host. Before we continue, I want to point out that thanks to the
```

File: _posts/2021-02-17-introduction-to-the-remote-execution-plugin.md [hunk 16]
```diff
@@ -0,0 +1,338 @@
...
+<%-   end -%>
+<%=   "uci commit" %>
+<%=   "/etc/init.d/led restart" %>
+<%- else -%>
+<%=   render_error('Unsupported OS for enabling LED') -%>
+<%- end -%>
+```
+
+Based on the state input it set the desired configuration. Then it commits the OpenWrt
+configuration and restarts the led service, which is the same code for both cases.
```
Comment (melcorr) [inline _posts/2021-02-17-introduction-to-the-remote-execution-plugin.md]:
```suggestion
configuration and restarts the LED service, which is the same code for both cases.
```

File: _posts/2021-02-17-introduction-to-the-remote-execution-plugin.md [hunk 17]
```diff
@@ -0,0 +1,338 @@
...
+Routers host group. The job template id is the id of the template I created.
+Similarly there’s a rule for disarming, just changing the state input value to off.
+The same flow of course is not limited to the openHAB, you can very easily convert
+this to curl command or fire the similar request form any other (home) automation tool.
+
+## Summary
+
+I hope you found this Remote Execution introduction useful and I’ll be happy if
+you share your experience and thoughts in the discussion below. I plan to incorporate
+REX in next posts too. The user interface for triggering the job is now being greatly
```
Comment (melcorr) [inline _posts/2021-02-17-introduction-to-the-remote-execution-plugin.md]:
```suggestion
REX in the next posts too. The user interface for triggering the job is now being greatly
```

**Review by melcorr — CHANGES_REQUESTED** [review]
Few nitpicks and can we s/puppet/Puppet please :) 
Thanks

**Comment by ares** [comment]
Thanks @melcorr, I've incorporated all your suggestions and capitalized puppet everywhere

**Pull request merged**